src/Psl/Hash/Context.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Hash;
+
+use HashContext;
+use Psl;
+use Psl\Arr;
+
+use function hash_final;
+use function hash_init;
+use function hash_update;
+
+/**
+ * Incremental hashing context.
+ *
+ * Example:
+ *
+ *      Hash\Context::create('md5')
+ *          ->update('The quick brown fox ')
+ *          ->update('jumped over the lazy dog.')
+ *          ->finalize()
+ *      => Str("5c6ffbdd40d9556b73a21e63c3e0e904")
+ *
+ * @psalm-immutable
+ */
+final class Context
+{
+    private HashContext $internalContext;
+
+    private function __construct(HashContext $internal_context)
+    {
+        $this->internalContext = $internal_context;
+    }
+
+    /**
+     * Initialize an incremental hashing context.
+     *
+     * @throws Psl\Exception\InvariantViolationException If the given algorithm is unsupported.
+     */
+    public static function create(string $algorithm): Context
+    {
+        Psl\invariant(Arr\contains(algorithms(), $algorithm), 'Unsupported algorithm "%s".', $algorithm);
+        $internal_context = hash_init($algorithm);
+
+        return new self($internal_context);
+    }
+
+    /**
+     * Pump data into an active hashing context.
+     */
+    public function update(string $data): Context
+    {
+        /** @var HashContext $internal_context */
+        $internal_context = hash_copy($this->internalContext);
```

Comment:
You are creating an immutable context class with this approach.
But I am wondering : does this come at a big memory cost if the data inside the internal context is quite big?

Reply:
if you are leaving a reference to the `Context` class somewhere, yes ... but if the object is desctructed, the internal context will be freed. meaning no issues.

i assume its mostly used as follow:

$context = Context::create('sha1');
foreach($block = $reader->read($block_size)) {
  $context = $context->update($block);
}

$hash = $context->finalize();
// there only 1 reference to `Context` left ( $context )